pyccel/ast/numpyext.py

```diff
@@ -342,14 +342,18 @@ def __init__(self, arg):
         if not isinstance(arg, PyccelAstNode):
             raise TypeError('Unknown type of  %s.' % type(arg))
         super().__init__(arg)
-        self._dtype = arg.dtype
+        self._arg = PythonList(arg) if arg.rank == 0 else self._args[0]
+        self._arg = PythonInt(self._arg) if (isinstance(arg.dtype, NativeBool) or \
+                    (isinstance(arg.dtype, NativeInteger) and self._arg.precision < default_precision['int']))\
+                    else self._arg
```

Comment:
I don't understand what is going on here in the case where `arg` is a scalar`bool`: first you convert it to a `list` and then you call `int()` on the list?

Reply:
it is not always 'then', if it is a scalar in general we should convert it to list because in Fortran `product` does not accept scalars, so the first block is just about scalar to list, the second one is about casting boolean to int and increase the precision of integers to the default_precision for the sake of a correct result.

Reply:
The confusion here, may be due to a modification made to `PythonInt` and the other casting functions.
When called on an array in python these functions will raise an error, but `PythonInt` will just cast the array to a different type

Reply:
I see. I did not remember that `PythonInt` could be applied to a `PythonList`.

Maybe we should update the docstrings of the casting functions...

Reply:
Actually, we dont mention which data structure we support in the docstring, but it needs to be more detailed.